packages/client/src/tlog/verify/checkpoint.ts

```diff
@@ -0,0 +1,231 @@
+  if (!signedNote.verify(validTLogs)) {
+    throw new VerificationError(
+      'inclusion proof contains invalid checkpoint signature'
+    );
+  }
+
+  // Verify that the root hash from the checkpoint matches the root hash in the
+  // inclusion proof
+  if (!crypto.bufferEqual(checkpoint.logHash, inclusionProof.rootHash)) {
+    throw new VerificationError(
+      'inclusion proof contains invalid root hash signature'
+    );
+  }
+}
+
+// SignedNote represents a signed note from a transparency log checkpoint. Consists
+// of a body (or note) and one more signatures calculated over the body. See
+// https://github.com/transparency-dev/formats/blob/main/log/README.md#signed-envelope
+class SignedNote {
+  readonly note: string;
+  readonly signatures: TLogSignature[];
+
+  constructor(note: string, signatures: TLogSignature[]) {
+    this.note = note;
+    this.signatures = signatures;
+  }
+
+  // Deserialize a SignedNote from a string
+  static fromString(envelope: string): SignedNote {
+    if (!envelope.includes(CHECKPOINT_SEPARATOR)) {
+      throw new VerificationError('malformed checkpoint: no separator');
+    }
+
+    // Split the note into the header and the data portions at the separator
+    const split = envelope.indexOf(CHECKPOINT_SEPARATOR);
+    const header = envelope.slice(0, split + 1);
+    const data = envelope.slice(split + CHECKPOINT_SEPARATOR.length);
+
+    // Find all the signature lines in the data portion
+    const matches = data.matchAll(SIGNATURE_REGEX);
+
+    // Parse each of the matched signature lines into the name and signature.
+    // The first four bytes of the signature are the key hint (should match the
+    // first four bytes of the log ID), and the rest is the signature itself.
+    const signatures = Array.from(matches, (match) => {
+      const [, name, signature] = match;
+      const sigBytes = Buffer.from(signature, 'base64');
+
+      if (sigBytes.length < 5) {
+        throw new VerificationError('malformed checkpoint: invalid signature');
+      }
+
+      return {
+        name,
+        keyHint: sigBytes.subarray(0, 4),
+        signature: sigBytes.subarray(4),
+      };
+    });
+
+    if (signatures.length === 0) {
+      throw new VerificationError('malformed checkpoint: no signatures');
+    }
+
+    return new SignedNote(header, signatures);
+  }
+
+  // Verifies the signatures in the SignedNote. For each signature, the
+  // corresponding transparency log is looked up by the key hint and the
+  // signature is verified against the public key in the transparency log.
+  // Throws an error if any of the signatures are invalid.
+  public verify(tlogs: ValidTransparencyLogInstance[]): boolean {
+    const data = Buffer.from(this.note, 'utf-8');
+
+    return this.signatures.every((signature) => {
+      // Find the transparency log instance with the matching key hint
+      const tlog = tlogs.find((tlog) =>
+        crypto.bufferEqual(tlog.logId.keyId.subarray(0, 4), signature.keyHint)
```

Comment:
potentially nit pick: should we recalculate the key id by hashing the public key? looks like they do this in sigstore-python but can't tell if this is overkill? https://github.com/sigstore/sigstore-python/blob/main/sigstore/_internal/rekor/checkpoint.py#L223